Add two new guides for Yjs and Tiptap (#2757)
liveblocks/liveblocks: guides/pages/yjs-best-practices-and-tips.mdx

---
meta:
  title: "Yjs best practices and tips"
  description:
    "Learn best practices and common pitfalls to avoid when working with Yjs and
    Liveblocks"
---

This guide covers best practices and common pitfalls to avoid when working with
Yjs and Liveblocks.

## Avoid importing Yjs twice

One of the most common issues when working with Yjs is accidentally importing it
twice in your application. This often happens when mixing CommonJS (CJS) and
ECMAScript Modules (ESM), or when certain bundlers bundle more than one version
of Yjs.

When Yjs is imported twice, the two instances don't share the same class
references, which can lead to synchronization issues and unexpected behavior.

> You will see a warning in the Console from Yjs if this happens.

Reviewer: ```suggestion
When Yjs is imported twice, the two instances don't share the same class
references, which can lead to synchronization issues and unexpected behavior. You'll see a Yjs warning in the Console if this happens.
```

updating alternating entries.

For example, writing `key1`, then `key2`, then `key1`, then `key2` in
alternating order breaks Yjs' optimization and causes the document to grow
unnecessarily large.

### Recommended approach: YKeyValue

For more efficient key-value storage, use
[`YKeyValue`](https://github.com/yjs/y-utility?tab=readme-ov-file#ykeyvalue)
from the `y-utility` package:

```bash
npm install y-utility
```

```ts
import * as Y from "yjs";
import { YKeyValue } from "y-utility/y-keyvalue";

const ydoc = new Y.Doc();
const yarr = ydoc.getArray();
const ykv = new YKeyValue(yarr);

// Fires events similarly to Y.Map when content changes
ykv.on("change", (changes) => {
  console.log(changes);
});

ykv.set("key1", "val1");
ykv.set("key1", "updated");
ykv.delete("key1");
ykv.set("key1", "new val");
ykv.get("key1"); // => 'new val'
```

`YKeyValue` creates documents whose size only depends on the size of the map,
not the number of operations. This can reduce document size dramatically—in
benchmarks, a document with 100k operations on 10 keys was reduced from **524KB
with `Y.Map`** to just **271 bytes with `YKeyValue`**.

> ## Enable experimental V2 encoding for LiveMaps

Reviewer: ```suggestion
## Enable experimental V2 encoding for Y.Maps
```

> If you're using `LiveMap` in combination with Yjs, you can enable the

Reviewer: ```suggestion
If you're using `Y.Map` in combination with Yjs, you can enable the
```

experimental V2 encoding for better performance and smaller document sizes:

```ts
import { useRoom } from "@liveblocks/react";
import { getYjsProviderForRoom } from "@liveblocks/yjs";

function App() {
  const room = useRoom();
  const yProvider = getYjsProviderForRoom(room, {
    // Enable V2 encoding for better performance with LiveMaps
    useV2Encoding_experimental: true,

>   });

Reviewer: ```suggestion

  const yProvider = getYjsProviderForRoom(room, {
    // Enable V2 encoding for better performance with LiveMaps
    // +++
    useV2Encoding_experimental: true,
    // +++
  });
```

}
```

> This encoding is more efficient when working with LiveMaps and can significantly

Reviewer: ```suggestion
This encoding is more efficient when working with maps and can significantly
```
